src/util/http/websocket/WebSocketManager.cpp

```diff
@@ -0,0 +1,189 @@
+                                                                  token);
+}
+
+net::awaitable<void> WebSocketManager::handleClientCommands(websocket& ws) {
+  beast::flat_buffer buffer;
+
+  while (ws.is_open()) {
+    co_await ws.async_read(buffer, net::use_awaitable);
+    // TODO<RobinTF> replace with cancellation process
+    ws.text(ws.got_text());
+    co_await ws.async_write(buffer.data(), net::use_awaitable);
+    buffer.clear();
+  }
+}
+
+net::awaitable<void> WebSocketManager::waitForServerEvents(
+    websocket& ws, const QueryId& queryId) {
+  UpdateFetcher updateFetcher{informationQueues_, wakeupCalls_, queryId,
+                              registryStrand_.value()};
+  while (ws.is_open()) {
+    auto json = co_await updateFetcher.waitForEvent(net::use_awaitable);
+    if (json == nullptr) {
+      if (ws.is_open()) {
+        co_await ws.async_close(beast::websocket::close_code::normal,
+                                net::use_awaitable);
+      }
+      break;
+    }
+    ws.text(true);
+    co_await ws.async_write(net::buffer(json->data(), json->length()),
+                            net::use_awaitable);
+  }
+}
+
+// Extracts the query id from a URL path. Returns the empty string when invalid.
+std::string extractQueryId(std::string_view path) {
+  auto match = ctre::match<"/watch/([^/?]+)">(path);
+  if (match) {
+    return match.get<1>().to_string();
+  }
+  return "";
+}
+
+net::awaitable<void> WebSocketManager::connectionLifecycle(
+    tcp::socket socket, http::request<http::string_body> request) {
+  auto queryIdString = extractQueryId(request.target());
+  AD_CORRECTNESS_CHECK(!queryIdString.empty());
+  QueryId queryId = QueryId::idFromString(queryIdString);
+  websocket ws{std::move(socket)};
+
+  try {
+    ws.set_option(beast::websocket::stream_base::timeout::suggested(
+        beast::role_type::server));
+
+    co_await ws.async_accept(request, boost::asio::use_awaitable);
+
+    auto strand = net::make_strand(ws.get_executor());
+
+    // experimental operators
+    co_await (
+        net::co_spawn(strand, waitForServerEvents(ws, queryId),
+                      net::use_awaitable) &&
+        net::co_spawn(strand, handleClientCommands(ws), net::use_awaitable));
+
+  } catch (boost::system::system_error& error) {
+    if (error.code() == beast::websocket::error::closed) {
+      co_return;
+    }
+    // There was an unexpected error, rethrow
+    throw;
+  }
+
+  // In case an unexpected exception is thrown, close the connection gracefully.
+  if (ws.is_open()) {
+    co_await ws.async_close(beast::websocket::close_code::internal_error,
+                            net::use_awaitable);
+  }
+}
+
+net::awaitable<void> WebSocketManager::manageConnection(
+    tcp::socket socket, http::request<http::string_body> request) {
+  auto executor = socket.get_executor();
+  return net::co_spawn(
+      executor, connectionLifecycle(std::move(socket), std::move(request)),
+      net::use_awaitable);
```

Comment:
I think the thread sanitizer only likes this, if the socket is either on a strand or on a dedicated `coro_executor` for reasons.